WireDomain/Sources/WireDomain/Services/UnknownMessageProcessingService.swift

```diff
@@ -0,0 +1,158 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import GenericMessageProtocol
+import WireDataModel
+import WireLogging
+
+/// Service responsible for processing stored unknown messages when the app is updated
+/// and new protobuf message types become available.
+public final class UnknownMessageProcessingService {
+
+    private let contextProvider: ContextProvider
+    private let conversationLocalStore: any ConversationLocalStoreProtocol
+    private let protobufMessageProcessor: any ConversationProtobufMessageProcessorProtocol
+    private let logger = WireLogger(tag: "UnknownMessageProcessing")
+
+    public init(
+        contextProvider: ContextProvider,
+        conversationLocalStore: any ConversationLocalStoreProtocol,
+        protobufMessageProcessor: any ConversationProtobufMessageProcessorProtocol
+    ) {
+        self.contextProvider = contextProvider
+        self.conversationLocalStore = conversationLocalStore
+        self.protobufMessageProcessor = protobufMessageProcessor
+    }
+
+    /// Processes all stored unknown messages by attempting to decode them with the current protobuf definitions.
+    /// Successfully processed messages are converted to their proper message types and the unknown message is deleted.
+    /// Messages that still cannot be decoded remain in the database for future processing.
+    public func processStoredUnknownMessages() async throws {
+        let context = contextProvider.syncContext
+
+        let unknownMessages = try await context.perform {
+            let fetchRequest = UnknownMessage.fetchRequest()
+            let unknownMessages = try context.fetch(fetchRequest)
+            return unknownMessages.map { ($0, $0.payload) }
+        }
+
+        logger.info("Found \(unknownMessages.count) stored unknown messages to process")
+
+        var processedCount = 0
+        var failedCount = 0
+
+        for (unknownMessage, payload) in unknownMessages {
+            do {
+                let wasProcessed = try await processUnknownMessage(unknownMessage, payload: payload)
+                if wasProcessed {
+                    processedCount += 1
+                } else {
+                    failedCount += 1
+                }
+            } catch {
+                logger.error("Failed to process unknown message: \(error)")
+                failedCount += 1
+            }
+        }
+
+        logger.info("Unknown message processing completed: \(processedCount) processed, \(failedCount) failed")
```

Comment:
question: when do we save the context? My expectation would be at the end of this method but I didn't give it too much thought.

Reply:
I injected a `databaseSaver` and added `save` at the end.